Give snap coloring a proper look and feel in Argon

mcendu opened the pull request:
Fixes #20977.

Before:

![osu_2024-05-10_22-53-28](https://github.com/ppy/osu/assets/43639171/2c223285-af7a-433a-b95e-8a5c28d3f5d8)

After:

![osu_2024-05-10_22-48-36](https://github.com/ppy/osu/assets/43639171/0d1d36e8-0113-4581-b66e-f2751945c028)

mcendu committed:
Add SnapDivisor bindable to DrawableManiaHitObject

This commit adds a new bindable, SnapDivisor, to DrawableManiaHitObject,
for consumption by skins. Logic for beat snap coloring is also moved
down.

Beat snap coloring is removed from Argon in the meantime, pending a new
design.

mcendu committed:
fix tests

peppy commented on osu.Game.Rulesets.Mania/Objects/Drawables/DrawableManiaHitObject.cs:
```diff
@@ -23,6 +23,11 @@ public abstract partial class DrawableManiaHitObject : DrawableHitObject<ManiaHi
+        /// <summary>
+        /// If timing based note colouring is used, the snap divisor of the note. 0 otherwise.
+        /// </summary>
+        public readonly Bindable<int> SnapDivisor = new Bindable<int>();
```
I'm not 100% on exposing the snap divisor here rather than the colour. There's precedent for exposing colours on the `DrawableHitObject` (`AccentColour`) after all.

@mcendu any reason for going in this direction instead?

mcendu replied:
To allow Argon to use its own colors for snap divisors, instead of using the editor snap color scheme for all skins.

peppy replied:
Hmm, I think at some point the colours would be defined by the skin, but the skin can still give the colours to the DHO rather than individual components.

Does argon already have custom colours somewhere? or is this for future direction?

peppy replied:
I think you might have missed my point. If we use accent colour, we wouldn't need the `SnapDivisor` bindable in the first place.

peppy commented:
I think this can go ahead without design work, at least to clean up the horrible state things are in currently.

mcendu commented:
> Does argon already have custom colours somewhere? or is this for future 
direction? For future direction.

peppy reviewed (commented):
Left some diff

mcendu committed:
properly implement snap colouring in Argon

Only notes and LN heads are colored. ppy made the initial
implementation; I tweaked the color scheme a bit to my own
preference. Flyte shall have the final say on this.

Co-Authored-By: peppy <pe@ppy.sh>

mcendu committed:
move snap update logic to base mania DHO

mcendu committed:
also snap color Argon hold note body and tail

This commit applies the head's snap color to Argon hold note bodies and
tails. For skin switching's sake the DHO's accent colour is not touched,
though the paradigm used causes quite a bit of code duplication here.

The look and feel is tentative programmer art, and arflyte shall (again)
have the final say on this.

mcendu committed:
fix code style

peppy commented on osu.Game.Rulesets.Mania/Objects/Drawables/DrawableManiaHitObject.cs:
```diff
@@ -41,7 +51,7 @@ protected override float SamplePlaybackPosition
         /// If non-null, judgements will be ignored whilst the function returns false.
         /// </summary>
-        public Func<DrawableHitObject, double, bool> CheckHittable;
+        public Func<DrawableHitObject, double, bool>? CheckHittable;
```
and stuf like this i dunno what has happened

mcendu replied:
Attempt to convert a single file to `#nullable true`
